### ISSUE TITLE
Vscode and atom highlight locations are unexpected when end location is not provided

### ISSUE THREAD

**Issue by golopot** [eslint/eslint#12334]
**Summary**
Many rules reports only start location and no end location. For these rules, editors often renders tildes in unexpected places.

**Examples**
In these examples, the rules only reports start location but no end locations. Instead of highlighting tokens like `\n`, `(`, or `{`, the token before the said token is highlighted. 

```js
/* eslint semi: [2, 'always'] */
const a = Math.random
//             ~~~~~~    actual highlight in vscode
//                   ~   expected highlight
//                   ^   reported start location


/* eslint object-curly-spacing: [2, 'always'] */
const b = {b: Math.random}
//                 ~~~~~~    actual highlight in vscode
//                       ~   expected highlight
//                       ^   reported start location


/* eslint space-in-parens: [2, 'never']  */
function ffff( a ) {}
//       ~~~~    actual highlight in vscode
//           ~   expected highlight
//           ^   reported start location
```
There are many other rules suffering from this problem, most of them are named `*space*` or `*spacing*`.

**How to fix**
The problem disappear if the both start location and end location are reported **and** start location is different to end location. That is, the location range is not zero-width-ed.

There are several places this can be done:

1. Modify each rules to report end location. Pros: explicit. Cons: lots of place needs change. 

2. Modify ESLint core, automatically add end location when end location is missing. Cons: not explicit.

3. Modify vscode-eslint plugin to automatically add end location when end location is missing.

**Comment by g-plane** [comment]
Maybe we can modify rules, because this can benefit formatters as well. However, I'm not sure how formatters behave when reporting those rules.

**Comment by kaicataldo** [comment]
I think one of the tricky things with some rules like semi is that there might not be a character to highlight if it's at the end of the line. Example from above:

```js
/* eslint semi: [2, 'always'] */
const a = Math.random
//                   ^ If there's no whitespace after the last character (and many editors can be configured
//                     to do this automatically), will VS Code be able to highlight it?
```

**Comment by golopot** [comment]
I have tested that case, report the location of the linefeed (width 1) will do well in vscode.

**Comment by kaicataldo** [comment]
Reopening this because this is actively being worked on.

**Comment by kaicataldo** [comment]
Leaving the "evaluating" label on here because we haven't officially accepted this, but have been evaluating and accepting smaller individual PRs that solve this.

**Comment by aladdin-add** [comment]
@eslint/eslint-team, it seems an useful enhancement, can we get one more 👍 to accept this issue?

**Comment by btmills** [comment]
Is the decision here to accept this issue as blanket acceptance for PRs to individual rules that add end locations to reports (option 1 in the original issue description) so we don't have to evaluate them individually? I've added my :+1: under that understanding, and if that is indeed the case, then this can be marked as accepted.

**Comment by kaicataldo** [comment]
@btmills I think that makes the most sense? Ideally, it would be nice to audit the rules and make a list we can keep track of here.

**Comment by golopot** [comment]
> ... Ideally, it would be nice to audit the rules and make a list we can keep track of here.

This is a list of rules that reports start column but misses end column. The list is acquired by patching rule tester in a dirty way.

- [ ] array-callback-return
- [ ] arrow-body-style
- [ ] block-spacing
- [ ] comma-style
- [ ] consistent-return
- [ ] curly
- [ ] eol-last
- [ ] func-call-spacing
- [ ] getter-return
- [ ] indent-legacy
- [ ] key-spacing
- [ ] keyword-spacing
- [ ] linebreak-style
- [ ] max-len
- [ ] max-lines
- [ ] new-cap
- [ ] newline-per-chained-call
- [ ] no-empty-function
- [ ] no-extra-bind
- [ ] no-irregular-whitespace
- [ ] no-mixed-spaces-and-tabs
- [ ] no-prototype-builtins
- [ ] no-redeclare
- [ ] no-spaced-func
- [ ] no-unexpected-multiline
- [ ] no-unneeded-ternary
- [ ] no-unused-vars
- [ ] no-useless-concat
- [ ] object-curly-newline
- [ ] object-property-newline
- [ ] one-var-declaration-per-line
- [ ] operator-linebreak
- [ ] padded-blocks
- [ ] rest-spread-spacing
- [ ] semi
- [ ] semi-spacing
- [ ] space-before-function-paren
- [ ] template-curly-spacing
- [ ] template-tag-spacing
- [ ] unicode-bom

**Comment by kaicataldo** [comment]
I will champion this. Given @btmills's  comment [here](https://github.com/eslint/eslint/issues/12334#issuecomment-554028203), I'm going to mark this as accepted. @golopot Thanks for spearheading this!

**Comment by mdjermanovic** [comment]
`semi` provides end location, except for the case where the start location is at the end of file.

Any ideas on how could we change this? Or maybe we should leave this in its current state?

There's a similar problem for `eol-last`.

**Comment by anikethsaha** [comment]
maybe line can be the same as `first.line` but the end column can be `first.column + 1` ?

**Comment by mdjermanovic** [comment]
> maybe line can be the same as `first.line` but the end column can be `first.column + 1` ?

Sorry, I didn't understand this.

For example, this case is problematic (there is no linebreak at the end of line 3)::

![image](https://user-images.githubusercontent.com/44349756/82137392-c0878c80-9817-11ea-95d6-43a9a27908fd.png)

In this case, `semi` reports the very last possible location as `start`, without `end`.

**Comment by anikethsaha** [comment]
I meant if the end can be like this 


```js
start: token.loc.start
end: {
  line: token.loc.start.line,
  column: token.loc.start.column + 1
}
```

**Comment by anikethsaha** [comment]
`keyword-spacing` has been fixed I suppose https://github.com/eslint/eslint/pull/13377. Can anyone update the list.

**Comment by mdjermanovic** [comment]
The remaining four rules may be problematic:

* `unicode-bom` - currently reports the very first location in the source code as the error's start loc. What could be the end loc?
* `consistent-return` (partially) - there's a similar problem as with `unicode-bom`, in the case where there's missing "global" return at the end of the program. Other cases can be fixed.
* `eol-last` - currently reports the very last location in the source code as the error's start loc. There are no valid locations after that location, so what could be the end loc?
* `semi` (partially) - there's a similar problem as with `eol-last`, in the case where the missing semi should be inserted after the EOF ([example in this comment](https://github.com/eslint/eslint/issues/12334#issuecomment-629751165)). All other cases are already fixed.

**Comment by nzakas** [comment]
I think for the harder cases we can either make a best guess at a change or just leave it as-is. We've already made so many fixes that I think we can be forgiven if a few aren't quite as accurate. `unicode-bom` and `eol-last` in particular, dealing with invisible characters, is something that is probably not going to make much sense so maybe just set the end location to one character after the start location?

**Comment by fisker** [comment]
Any interest to implement a visualize tester like [this](https://github.com/sindresorhus/eslint-plugin-unicorn/pull/788/files#diff-173dbb65c310faa266fd4d28870f7c04R1) to test locations? Or even better add to `RuleTester`?

**Comment by anikethsaha** [comment]
Do you mean snapshot testing ?

**Comment by fisker** [comment]
I mean this code frame

```markdown
    `␊
      1 | function foo() {␊
    > 2 | 	function bar() {}␊
        | 	^^^^^^^^^^^^ Move function 'bar' to the outer scope.␊
      3 | }␊
    `
```

Easier to test.

**Comment by anikethsaha** [comment]
I meant, how would we assert that ?  I can think of snapshot testing only other than any hacky approach.
Or you meant to do the assertion like it is currently setup and just visually display the frame ?

**Comment by fisker** [comment]
Yes, snapshot should be the easiest way.

**Comment by nzakas** [comment]
@fisker I’d suggest opening a separate issue to discuss your idea.

**Comment by snitin315** [comment]
Taking this up.

**Comment by snitin315** [comment]
This is all done 🎉 .

**Comment by nzakas** [comment]
Awesome! Thanks to everyone who helped push through this. 🙏
